File: backend/src/llm_followups/rules/engine.py
from __future__ import annotations

from collections.abc import Sequence

from llm_followups.server.schemas import ChatMessage

from .definitions import DEFAULT_RULES
from .matching import contains_keyword
from .models import (
    ConversationRule,
    RuleStep,
)


NO_RULE_RESPONSE = (
    "I don't have a predefined rule for that topic yet."
)


class RuleEngine:
    """
    Deterministic conversation engine.

    The engine never calls an LLM. Flow state is reconstructed from
    persisted conversation history, so a backend restart does not lose
    the active rule step.
    """

    def __init__(
        self,
        rules: Sequence[ConversationRule] = DEFAULT_RULES,
    ) -> None:
        self._rules = tuple(rules)

        self._question_index: dict[
            str,
            tuple[ConversationRule, int],
        ] = {}

        for rule in self._rules:
            for index, step in enumerate(rule.steps):
                if step.question in self._question_index:
                    raise ValueError(
                        "Rule questions must be unique: "
                        f"{step.question}"
                    )

                self._question_index[
                    step.question
                ] = (rule, index)

    def match_rule(
        self,
        text: str,
    ) -> ConversationRule | None:
        """
        Return the most specific matching topic rule.

        More matched keywords wins. Definition order is the stable
        tie-breaker.
        """

        best_rule: ConversationRule | None = None
        best_score = 0

        for rule in self._rules:
            score = sum(
                1
                for keyword in rule.keywords
                if contains_keyword(text, keyword)
            )

            if score > best_score:
                best_rule = rule
                best_score = score

        return best_rule

    def _pending_step(
        self,
        messages: Sequence[ChatMessage],
    ) -> tuple[ConversationRule, int] | None:
        """
        Find the latest predefined question in assistant history.

        The first assistant message encountered while scanning backward
        is authoritative. If it contains no rule question, the prior
        flow is considered complete.
        """

        for message in reversed(messages):
            if message.role == "user":
                continue

            content = message.content

            for question, location in self._question_index.items():
                if question in content:
                    return location

            return None

        return None

    @staticmethod
    def _matching_branch(
        step: RuleStep,
        user_answer: str,
    ):
        for branch in step.branches:
            if any(
                contains_keyword(
                    user_answer,
                    keyword,
                )
                for keyword in branch.keywords
            ):
                return branch

        return None

    def _resolve_step_answer(
        self,
        step: RuleStep,
        user_answer: str,
    ) -> str:
        branch = self._matching_branch(
            step,
            user_answer,
        )

        if branch is not None:
            return branch.response

        return step.default_response

    def _topic_switch(
        self,
        *,
        current_rule: ConversationRule,
        current_step: RuleStep,
        user_text: str,
    ) -> ConversationRule | None:
        """
        Detect an intentional topic change while a rule question is
        pending.

        A reply that already matches one of the current step's answer
        branches is treated as an answer, even if it also contains some
        other topic keyword. Otherwise, a clear match for a different
        rule starts that rule immediately.
        """

        if (
            self._matching_branch(
                current_step,
                user_text,
            )
            is not None
        ):
            return None

        candidate = self.match_rule(
            user_text
        )

        if (
            candidate is None
            or candidate.id == current_rule.id
        ):
            return None

        return candidate

    def respond(
        self,
        messages: Sequence[ChatMessage],
    ) -> str:
        """
        Produce the next deterministic response.

        Initial topic:
            -> one predefined question

        Rule answer:
            -> predefined answer
            -> optionally one next predefined question

        Topic switch:
            -> first question for the newly matched rule

        Completed/unmatched topic:
            -> deterministic no-rule response
        """

        if not messages:
            raise ValueError(
                "At least one chat message is required"
            )

        latest = messages[-1]

        if latest.role != "user":
            raise ValueError(
                "The latest chat message must be from the user"
            )

        prior_messages = messages[:-1]

        pending = self._pending_step(
            prior_messages
        )

        if pending is not None:
            rule, step_index = pending
            step = rule.steps[step_index]

            switched_rule = self._topic_switch(
                current_rule=rule,
                current_step=step,
                user_text=latest.content,
            )

            if switched_rule is not None:
                if not switched_rule.steps:
                    return NO_RULE_RESPONSE

                return (
                    switched_rule
                    .steps[0]
                    .question
                )

            answer = self._resolve_step_answer(
                step,
                latest.content,
            )

            next_index = step_index + 1

            if next_index < len(rule.steps):
                next_question = (
                    rule.steps[next_index]
                    .question
                )

                return (
                    f"{answer}\n\n"
                    f"{next_question}"
                )

            return answer

        rule = self.match_rule(
            latest.content
        )

        if rule is None or not rule.steps:
            return NO_RULE_RESPONSE

        return rule.steps[0].question

File: frontend/src/rules/engine.ts
import type {
  ChatMessage,
} from "../types/chat";

import type {
  AnswerBranch,
  ConversationRule,
  RuleStep,
} from "./models";

export const NO_RULE_RESPONSE =
  "I don't have a predefined rule for that topic yet.";

function normalise(
  text: string,
): string {
  return text
    .toLocaleLowerCase()
    .trim()
    .replace(/\s+/g, " ");
}

function escapeRegExp(
  value: string,
): string {
  return value.replace(
    /[.*+?^${}()|[\]\\]/g,
    "\\$&",
  );
}

function containsKeyword(
  text: string,
  keyword: string,
): boolean {
  const normalisedText =
    normalise(text);

  const normalisedKeyword =
    normalise(keyword);

  if (!normalisedKeyword) {
    return false;
  }

  if (
    normalisedKeyword.includes(" ") ||
    normalisedKeyword.includes("-")
  ) {
    return normalisedText.includes(
      normalisedKeyword,
    );
  }

  const pattern =
    new RegExp(
      `(^|\\W)${escapeRegExp(
        normalisedKeyword,
      )}(?=\\W|$)`,
      "i",
    );

  return pattern.test(
    normalisedText,
  );
}

interface PendingStep {
  rule: ConversationRule;
  stepIndex: number;
}

export class RuleEngine {
  private readonly rules:
    readonly ConversationRule[];

  private readonly questionIndex =
    new Map<
      string,
      PendingStep
    >();

  constructor(
    rules: readonly ConversationRule[],
  ) {
    this.rules = rules;

    for (const rule of rules) {
      rule.steps.forEach(
        (step, stepIndex) => {
          if (
            this.questionIndex.has(
              step.question,
            )
          ) {
            throw new Error(
              `Rule questions must be unique: ${step.question}`,
            );
          }

          this.questionIndex.set(
            step.question,
            {
              rule,
              stepIndex,
            },
          );
        },
      );
    }
  }

  matchRule(
    text: string,
  ): ConversationRule | null {
    let bestRule:
      ConversationRule | null =
      null;

    let bestScore = 0;

    for (const rule of this.rules) {
      const score =
        rule.keywords.reduce(
          (total, keyword) =>
            total +
            (
              containsKeyword(
                text,
                keyword,
              )
                ? 1
                : 0
            ),
          0,
        );

      if (score > bestScore) {
        bestRule = rule;
        bestScore = score;
      }
    }

    return bestRule;
  }

  private pendingStep(
    messages: readonly ChatMessage[],
  ): PendingStep | null {
    for (
      let index =
        messages.length - 1;
      index >= 0;
      index -= 1
    ) {
      const message =
        messages[index];

      if (
        message.role === "user"
      ) {
        continue;
      }

      for (
        const [
          question,
          location,
        ] of
        this.questionIndex.entries()
      ) {
        if (
          message.content.includes(
            question,
          )
        ) {
          return location;
        }
      }

      return null;
    }

    return null;
  }

  private matchingBranch(
    step: RuleStep,
    userAnswer: string,
  ): AnswerBranch | null {
    for (
      const branch
      of step.branches
    ) {
      if (
        branch.keywords.some(
          (keyword) =>
            containsKeyword(
              userAnswer,
              keyword,
            ),
        )
      ) {
        return branch;
      }
    }

    return null;
  }

  private resolveStepAnswer(
    step: RuleStep,
    userAnswer: string,
  ): string {
    return (
      this.matchingBranch(
        step,
        userAnswer,
      )?.response ??
      step.defaultResponse
    );
  }

  private topicSwitch(
    currentRule:
      ConversationRule,
    currentStep: RuleStep,
    userText: string,
  ): ConversationRule | null {
    if (
      this.matchingBranch(
        currentStep,
        userText,
      )
    ) {
      return null;
    }

    const candidate =
      this.matchRule(userText);

    if (
      !candidate ||
      candidate.id ===
        currentRule.id
    ) {
      return null;
    }

    return candidate;
  }

  respond(
    messages:
      readonly ChatMessage[],
  ): string {
    if (
      messages.length === 0
    ) {
      throw new Error(
        "At least one chat message is required.",
      );
    }

    const latest =
      messages[
        messages.length - 1
      ];

    if (
      latest.role !== "user"
    ) {
      throw new Error(
        "The latest chat message must be from the user.",
      );
    }

    const priorMessages =
      messages.slice(0, -1);

    const pending =
      this.pendingStep(
        priorMessages,
      );

    if (pending) {
      const {
        rule,
        stepIndex,
      } = pending;

      const step =
        rule.steps[
          stepIndex
        ];

      const switchedRule =
        this.topicSwitch(
          rule,
          step,
          latest.content,
        );

      if (switchedRule) {
        return (
          switchedRule.steps[0]
            ?.question ??
          NO_RULE_RESPONSE
        );
      }

      const answer =
        this.resolveStepAnswer(
          step,
          latest.content,
        );

      const nextStep =
        rule.steps[
          stepIndex + 1
        ];

      if (nextStep) {
        return (
          `${answer}\n\n` +
          nextStep.question
        );
      }

      return answer;
    }

    const rule =
      this.matchRule(
        latest.content,
      );

    return (
      rule?.steps[0]
        ?.question ??
      NO_RULE_RESPONSE
    );
  }
}
